### PR TITLE
periphery: convert bundle traits 

### PR DESCRIPTION
**Pull request by hcook**
...to work with system-level multi-io module.

The abstract bundle traits can either be mixed into Bundles and made concrete with calls to `cloneType`, or mixed into ModuleImps and made concrete with `IO(...)`s. Examples of the latter use case are provided for each peripheral.

TODO in future PR: add more helper functions in the style of `toGPIOPins`

### CODE REVIEW

File: src/main/scala/devices/i2c/I2CPeriphery.scala [hunk 1]
```diff
@@ -19,15 +19,21 @@ trait HasPeripheryI2C extends HasTopLevelNetworks {
   }
 }
 
-trait HasPeripheryI2CBundle extends HasTopLevelNetworksBundle{
-  val outer: HasPeripheryI2C
-  val i2cs = Vec(outer.i2cParams.size, new I2CPort)
+trait HasPeripheryI2CBundle {
+  val i2cs: Vec[I2CPort]
+
+  def toGPIOPins(dummy: Int = 1): Seq[I2CGPIOPort] = i2cs.map { i =>
```
Comment (mwachs5) [inline src/main/scala/devices/i2c/I2CPeriphery.scala]: what is the (dummy) for? I think we will need the synchronizers at least. Maybe we should make the synchronizer depth part of each peripheral parameter.

File: src/main/scala/devices/spi/SPIPeriphery.scala [hunk 2]
```diff
@@ -24,22 +20,28 @@ trait HasPeripherySPI extends HasTopLevelNetworks {
...
+    pin
+  }
 }
 
-trait HasPeripherySPIModule extends HasTopLevelNetworksModule {
+trait HasPeripherySPIModuleImp extends LazyMultiIOModuleImp with HasPeripherySPIBundle {
   val outer: HasPeripherySPI
-  val io: HasPeripherySPIBundle
-  (io.spis zip outer.spis).foreach { case (io, device) =>
+  val spis = IO(HeterogeneousBag(outer.spiParams.map(new SPIPortIO(_))))
```
Comment (mwachs5) [inline src/main/scala/devices/spi/SPIPeriphery.scala]: I realized I was inconsistent in the naming of "spis" vs "qspi", etc. We should probably standardize (could be in a different PR)

File: src/main/scala/devices/i2c/I2CPeriphery.scala (hunk 1 above)
Comment (hcook) [inline src/main/scala/devices/i2c/I2CPeriphery.scala]: Members of chisel bundles that don't take parameters are automatically converted into wires. So this is the standard gross hack to avoid that from happening before you want it to. Happy to replace it with a parameter that is actually useful.

Comment (mwachs5) [inline src/main/scala/devices/i2c/I2CPeriphery.scala]: Ah, i see.  Well any of the "GPIOPort" which has inputs should allow a synchronizer depth parameter. I think right now only SPIGPIOPort has that.

File: src/main/scala/devices/spi/SPIPeriphery.scala (hunk 2 above)
Comment (hcook) [inline src/main/scala/devices/spi/SPIPeriphery.scala]: Yeah we definitely aren't consistent about pluralization. i noticed @henrystyles added an 's to a peripheral io recently. I don't really have a strong opinion either way

**Comment by mwachs5** [comment]
What do you expect to happen for the clocking for some of these mixins? There are registers instantiated inside  *toGPIOPins. Do you expect all these methods to be called from inside withClock(...) blocks? I guess that is no worse than the current situation where you have to make sure to assign the .clock correctly for the *GPIOPort.

**Comment by hcook** [comment]
It is a goal to start thinking about the multi-clock stuff as well. If we imagine a design where each peripheral is taking a unique clock as an input (possibly from the top level), we want to use the same clock to drive the pin registers / test SRAMs / blackboxed VIP attached to that port. Unfortunately i don't think we want to pipe the clock back out as an output, though that would make it easy to hook up automatically. For now i would just have the helper functions take clock and/or reset as an explicit argument, and we can wrap them up more intelligently later

**Comment by mwachs5** [comment]
That makes sense, but I'm not planning to do it (passing the clock in) for now.
